### PR TITLE
feat: add failed key path and line to output

### PR DESCRIPTION
**Pull request by shmu3l**

### CODE REVIEW

**Comment by hadar-co** [comment]
Screen Shot 2022-08-30 at 16 04 32

File: pkg/cliClient/evaluation.go [hunk 1]
```diff
@@ -155,12 +155,19 @@ type FileData struct {
 	ConfigurationsCount int    `json:"configurationsCount"`
 }
 
+type ValidationResult struct {
```
Comment (royhadad) [inline pkg/cliClient/evaluation.go]: rename to `FailureLocation`?

File: pkg/extractor/extractor.go [hunk 2]
```diff
@@ -144,7 +145,7 @@ func extractConfigurationK8sData(content []byte) Configuration {
 	if kind := jsonObject["kind"]; kind != nil {
 		configuration.Kind = kind.(string)
 	}
-
+	configuration.YamlNode = yamlNode
```
Comment (royhadad) [inline pkg/extractor/extractor.go:151]: should be lower case `configuration.yamlNode`

File: pkg/printer/printer_test.go [hunk 3]
```diff
@@ -14,10 +14,16 @@ func TestGetWarningsText(t *testing.T) {
...
-				Suggestion:         "Suggestion",
-				OccurrencesDetails: []OccurrenceDetails{{MetadataName: "yishay", Kind: "Pod"}},
+				Name:        "Caption",
+				Occurrences: 1,
+				Suggestion:  "Suggestion",
+				OccurrencesDetails: []OccurrenceDetails{{
+					MetadataName:      "yishay",
+					Kind:              "Pod",
+					FailedErrorLine:   0,
+					FailedErrorColumn: 0,
```
Comment (royhadad) [inline pkg/printer/printer_test.go]: make line and column different, this way the test case will catch more cases, when someone accidentally flips the printing order. (better value for money in the same test

File: pkg/printer/printer_test.go [hunk 4]
```diff
@@ -56,6 +62,8 @@ func TestGetWarningsText(t *testing.T) {
 
 ❌  Caption  [1 occurrence]
     - metadata.name: yishay (kind: Pod)
+      > key: .spec.containers[0].name (line: 0:0)
+
```
Comment (royhadad) [inline pkg/printer/printer_test.go:69]: do we want the extra line break here? it wasn't here before (this is a product question)

File: pkg/evaluation/evaluator.go [hunk 5]
```diff
@@ -362,3 +380,29 @@ func extractSkipAnnotations(configuration extractor.Configuration) map[string]st
 
 	return skipAnnotations
 }
+
+func updateFailedRuleLine(schemaPath string, yamlNode yaml.Node) (failedErrorLine int, failedErrorColumn int) {
```
Comment (royhadad) [inline pkg/evaluation/evaluator.go]: rename method `get...` :)

File: pkg/evaluation/evaluator.go [hunk 6]
```diff
@@ -362,3 +380,29 @@ func extractSkipAnnotations(configuration extractor.Configuration) map[string]st
...
+	if err != nil {
+		return
+	}
+
+	candidateNode := nodeList.Back().Value.(*yqlib.CandidateNode).Node
+
+	return candidateNode.Line, candidateNode.Column
+}
+
+func yqlibLoggerHandler() {
```
Comment (royhadad) [inline pkg/evaluation/evaluator.go]: What is this? please explain it to me face2face :)

File: pkg/evaluation/evaluator.go (hunk 5 above)
Comment (royhadad) [inline pkg/evaluation/evaluator.go]: now that this method is **pure** we can (and should) add a unit test for it, there is some complex logic here

File: pkg/evaluation/evaluator.go [hunk 7]
```diff
@@ -362,3 +380,29 @@ func extractSkipAnnotations(configuration extractor.Configuration) map[string]st
 
 	return skipAnnotations
 }
+
+func updateFailedRuleLine(schemaPath string, yamlNode yaml.Node) (failedErrorLine int, failedErrorColumn int) {
+
+	instanceLocationYqPath := strings.Replace(schemaPath, "/", ".", -1)
```
Comment (royhadad) [inline pkg/evaluation/evaluator.go:399]: please explain this method to me face2face :)

**Review by royhadad — CHANGES_REQUESTED** [review]

File: pkg/evaluation/evaluator_test.go [hunk 8]
```diff
@@ -163,6 +165,20 @@ func TestEvaluate(t *testing.T) {
 	}
 }
 
+func TestGetFailedRuleLineAndColumn(t *testing.T) {
+	yamlNode := yaml.Node{
```
Comment (royhadad) [inline pkg/evaluation/evaluator_test.go]: embed a real yaml and convert it to `yaml.Node`

File: pkg/cliClient/evaluation.go [hunk 9]
```diff
@@ -155,12 +155,19 @@ type FileData struct {
...
-	Kind        string `json:"kind"`
-	Occurrences int    `json:"occurrences"`
-	IsSkipped   bool   `json:"isSkipped"`
-	SkipMessage string `json:"skipMessage"`
+	Name             string            `json:"metadataName"`
+	Kind             string            `json:"kind"`
+	Occurrences      int               `json:"occurrences"`
+	IsSkipped        bool              `json:"isSkipped"`
+	SkipMessage      string            `json:"skipMessage"`
+	FailureLocations []FailureLocation `json:"FailureLocation"`
```
Comment (myishay) [inline pkg/cliClient/evaluation.go]: json tag should be lowercased

File: pkg/printer/printer.go [hunk 10]
```diff
@@ -207,6 +209,12 @@ func (p *Printer) GetWarningsText(warnings []Warning) string {
 
 				for _, occurrenceDetails := range failedRule.OccurrencesDetails {
 					sb.WriteString(fmt.Sprintf("    - metadata.name: %v (kind: %v)\n", p.getStringOrNotAvailableText(occurrenceDetails.MetadataName), p.getStringOrNotAvailableText(occurrenceDetails.Kind)))
+					for _, validationResult := range occurrenceDetails.FailureLocations {
+						errorPath := fmt.Sprintf("%v (line: %d:%d)\n", strings.Replace(validationResult.SchemaPath, "/", ".", -1)[1:], validationResult.FailedErrorLine, validationResult.FailedErrorColumn)
```
Comment (myishay) [inline pkg/printer/printer.go]: ins this an error? or a failure of the rule?

**Review by myishay — DISMISSED** [review]

**Review by royhadad — APPROVED** [review]

**Pull request merged**